lib/ansible/galaxy/role.py

```diff
@@ -393,14 +443,16 @@ def install(self):
                             # bits that might be in the file for security purposes
                             # and drop any containing directory, as mentioned above
                             if member.isreg() or member.issym():
-                                for attr in ('name', 'linkname'):
+                                for attr in ("name", "linkname"):
                                     attr_value = getattr(member, attr, None)
                                     if not attr_value:
                                         continue
                                     n_attr_value = to_native(attr_value)
                                     n_archive_parent_dir = to_native(archive_parent_dir)
-                                    n_parts = n_attr_value.replace(n_archive_parent_dir, "", 1).split(os.sep)
-                                    n_final_parts = []
+                                    n_parts = n_attr_value.replace(
+                                        n_archive_parent_dir, "", 1
+                                    ).split(os.sep)
+                                    n_final_parts = [""]
```

Comment:
This ends up installing a broken role (no tests/roles/ansible-sshd/ dir). It looks like content in the role https://github.com/willshersystems/ansible-sshd/tree/main/tests/roles/ansible-sshd is relying on symlinks, which was restricted in https://github.com/ansible/ansible/pull/81780 and backported to 2.13.

Should the error be improved instead? If we do allow partial-installs, the warning could be made less cryptic, here's a more minimal patch:

```diff
--- a/lib/ansible/galaxy/role.py
+++ b/lib/ansible/galaxy/role.py
@@ -390,16 +390,17 @@ class GalaxyRole(object):
                                         # Check if we have any files with illegal names,
                                         # and display a warning if so. This could help users
                                         # to debug a broken installation.
+                                        warning_context = f"Invalid {attr} for member {member.name} ({attr_value}). This indicates an issue with the role, and you may end up with a broken install."
                                         if not n_part :
-                                            continue
+                                            pass
                                         if n_part == '..':
-                                            display.warning(f"Illegal filename '{n_part}': '..' is not allowed")
+                                            display.warning(f"{warning_context} Illegal filename '{n_part}': '..' is not allowed")
                                             continue
                                         if n_part.startswith('~'):
-                                            display.warning(f"Illegal filename '{n_part}': names cannot start with '~'")
+                                            display.warning(f"{warning_hint} Illegal filename '{n_part}': names cannot start with '~'")
                                             continue
                                         if '$' in n_part:
-                                            display.warning(f"Illegal filename '{n_part}': names cannot contain '$'")
+                                            display.warning(f"{warning_hint} Illegal filename '{n_part}': names cannot contain '$'")
                                             continue
                                         n_final_parts.append(n_part)
                                     setattr(member, attr, os.path.join(*n_final_parts))
```

Results in:
```
...
[WARNING]: Invalid linkname for member tests/roles/ansible-sshd (../../). This indicates an issue with the role, and you may end up with a broken install. Illegal filename '..': '..' is not allowed
- willshersystems.sshd (v0.21.0) was installed successfully
```

Reply:
-10 to 'partial install' .. we should either succeed or not, leaving a partially installed role could cause all kinds of random issues. i.e installed tasks but not handlers/vars/defaults/plugins they use or install modules but not their module_utils, ...

Reply:
At the moment `ansible-galaxy install` crashes in the middle of the install, which is resulting in a partial install. This is sure to break things and it was quite hard to figure out.

So IMHO partial installs should be avoided.

A concensus should be reached about how to deal with this and most likely a new pull request should be raised.

My code-formatter jumped in on my request, that's why there are a couple of unrelated formatting changes.
Apologies for having missed that.

As the final solution will most likely be different to what I initially suggested, IMHO it doesn't make sense for me to update my pull request.

Reply:
BTW: Meanwhile the erroneous symlink was removed in [wilshersystems/ansible-sshd](https://github.com/willshersystems/ansible-sshd/pull/262)